scripts/cairo_programs_verifier.sh

```diff
@@ -4,39 +4,67 @@ GREEN='\033[1;32m'
 RED='\033[1;31m'
 NC='\033[0m'
 
-# flag to check if any errors were encountered
-has_errors=false
-
-# function to process directory
-process_directory() {
-  for dir in "$1"/*
-  do
-    if [ -d "$dir" ]; then
-      echo "Processing $dir"
-      (cd "$dir" && scarb build 0>/dev/null 1> error.log && scarb fmt -c 0>/dev/null 1>> error.log && scarb test 0>/dev/null 1>> error.log)
-      if [ $? -ne 0 ]; then
-        has_errors=true
-        echo "Error while processing $dir"
-        cat "$dir/error.log"
-      fi
-      rm "$dir/error.log"
+# root directory of the repository
+REPO_ROOT="$(git rev-parse --show-toplevel)"
+error_file=$(mktemp)
+
+# function to list modified cairo files
+list_modified_cairo_files() {
+    git diff --name-only main...HEAD -- listings | grep -E 'listings/ch.*/*.cairo$'
+}
+
+# function to process individual file
+process_file() {
+    local relative_path="$1"
+    local dir_path="${REPO_ROOT}/$(dirname "${relative_path}")"
+    local file_name=$(basename "$relative_path")
+
+    if ! cd "$dir_path"; then
+        echo "Failed to change to directory: $dir_path"
+        echo "1" >> "$error_file"
+        return
     fi
-  done
+
+    if ! scarb build >error.log 2>&1; then
+        cat error.log
+        echo "1" >> "$error_file"
+    fi
+
+    if ! scarb fmt >>error.log 2>&1; then
+        echo "Error in scarb format check for $file_name"
+        cat error.log
+        echo "1" >> "$error_file"
+    fi
+
+    if ! scarb test >>error.log 2>&1; then
+        echo "Error in scarb test for $file_name"
+        cat error.log
+        echo "1" >> "$error_file"
+    fi
+
+    rm error.log
 }
 
-# only start processing directories that match the pattern ch*-* 
-for dir in listings/ch*-*
-do
-  if [ -d "$dir" ]; then
-    process_directory "$dir"
-  fi
+# process each modified file
+pids=()
+modified_files=$(list_modified_cairo_files)
+for file in $modified_files; do
+    process_file "$file" &
+    pids+=($!)
+done
+
+# wait for all background processes to finish
+for pid in ${pids[@]}; do
+    wait $pid
 done
 
 # check if any errors were encountered
-if $has_errors ; then
-  echo "\n${RED}Some projects have errors, please check the list above.${NC}\n"
-  exit 1
+if grep -q "1" "$error_file"; then
+    echo -e "\n${RED}Some projects have errors, please check the list above.${NC}\n"
+    rm "$error_file"
+    exit 1
 else
-  echo -e "\n${GREEN}All scarb builds were completed successfully${NC}.\n"
-  exit 0
+    echo -e "\n${GREEN}All scarb builds were completed successfully${NC}.\n"
```

Comment:
Suggestion: count the number of detected changed listing and print `All {listings_count} buiilds were completed successfully` and if no change `No new changes detected.`